pallets/restricted-xtokens/src/tests.rs

```diff
@@ -0,0 +1,538 @@
+		TestNet::reset();
+
+		Relay::execute_with(|| {
+			let _ = RelayBalances::deposit_creating(&para_a_account(), 1_000);
+		});
+
+		ParaA::execute_with(|| {
+			assert_noop!(
+				ParaXTokens::transfer(
+					Some(RESTRICTED_SENDER).into(),
+					CurrencyId::R,
+					500,
+					Box::new(
+						MultiLocation::new(
+							1,
+							X1(Junction::AccountId32 {
+								network: None,
+								id: BOB.into(),
+							})
+						)
+						.into()
+					),
+					WeightLimit::Unlimited
+				),
+				Error::<para::Runtime>::RestrictionTriggered
+			);
+
+			assert_eq!(
+				ParaTokens::free_balance(CurrencyId::R, &RESTRICTED_SENDER),
+				1000
+			);
+		});
+
+		Relay::execute_with(|| {
+			assert_eq!(RelayBalances::free_balance(&para_a_account()), 1000);
+			assert_eq!(RelayBalances::free_balance(&BOB), 0);
+		});
+	}
+
+	#[test]
+	fn allow_normal_transfer_relay_currency() {
+		TestNet::reset();
+
+		Relay::execute_with(|| {
+			let _ = RelayBalances::deposit_creating(&para_a_account(), 1_000);
+		});
+
+		ParaA::execute_with(|| {
+			assert_ok!(ParaXTokens::transfer(
+				Some(RESTRICTED_SENDER).into(),
+				CurrencyId::R,
+				500,
+				Box::new(
+					MultiLocation::new(
+						1,
+						X1(Junction::AccountId32 {
+							network: None,
+							id: RESTRICTED_RECEIVER.into(),
+						})
+					)
+					.into()
+				),
+				WeightLimit::Unlimited
+			));
+			assert_eq!(
+				ParaTokens::free_balance(CurrencyId::R, &RESTRICTED_SENDER),
+				500
+			);
+		});
+
+		Relay::execute_with(|| {
+			assert_eq!(RelayBalances::free_balance(&para_a_account()), 500);
+			assert_eq!(RelayBalances::free_balance(&RESTRICTED_RECEIVER), 460);
+		});
+	}
+}
+
+mod para_a_to_para_a {
+	use super::*;
+
+	#[test]
+	fn restrict_normal_transfer_relay_currency() {
+		TestNet::reset();
+
+		ParaA::execute_with(|| {
+			assert_noop!(
+				ParaXTokens::transfer(
+					Some(RESTRICTED_SENDER).into(),
+					CurrencyId::R,
+					500,
+					Box::new(
+						MultiLocation::new(
+							0,
+							X1(Junction::AccountId32 {
+								network: None,
+								id: BOB.into(),
+							})
+						)
+						.into()
+					),
+					WeightLimit::Unlimited
+				),
+				Error::<para::Runtime>::RestrictionTriggered
+			);
+
+			assert_eq!(
+				ParaTokens::free_balance(CurrencyId::R, &RESTRICTED_SENDER),
+				1000
+			);
+		});
+	}
+
+	#[test]
+	fn restrict_normal_transfer_a1_currency() {
+		TestNet::reset();
+
+		ParaA::execute_with(|| {
+			assert_noop!(
+				ParaXTokens::transfer(
+					Some(RESTRICTED_SENDER).into(),
+					CurrencyId::A1,
+					500,
+					Box::new(
+						MultiLocation::new(
+							0,
+							X1(Junction::AccountId32 {
+								network: None,
+								id: BOB.into(),
+							})
+						)
+						.into()
+					),
+					WeightLimit::Unlimited
+				),
+				Error::<para::Runtime>::RestrictionTriggered
+			);
+
+			assert_eq!(
+				ParaTokens::free_balance(CurrencyId::R, &RESTRICTED_SENDER),
+				1000
+			);
+		});
+	}
+
+	#[test]
+	fn restrict_normal_transfer_b1_currency() {
```

Comment:
I will keep reading but I wonder here what a "restricted normal transfer" means

Reply:
normal is just the `transfer()` method.